benchmark/Bench/State.hs

```diff
@@ -0,0 +1,36 @@
+module Bench.State
+( benchmark
+) where
+
+import Control.Carrier.State.Church as C.Church
+import Control.Carrier.State.Lazy as C.Lazy
+import Control.Carrier.State.Strict as C.Strict
+import Control.Monad.Trans.State.Lazy as T.Lazy (execStateT)
+import Control.Monad.Trans.State.Strict as T.Strict (execStateT)
+import Data.Foldable (for_)
+import Gauge hiding (benchmark)
+
+benchmark :: Benchmark
+benchmark = bgroup "State"
+  [ bgroup "Identity"
+    [ bench "Church.StateC" $ whnf (run . C.Church.execState from . modLoop) n
+    , bench "Lazy.StateC"   $ whnf (run . C.Lazy.execState from . modLoop) n
+    , bench "Strict.StateC" $ whnf (run . C.Strict.execState from . modLoop) n
+    , bench "Lazy.StateT"   $ whnf (run . flip T.Lazy.execStateT from . modLoop) n
+    , bench "Strict.StateT" $ whnf (run . flip T.Strict.execStateT from . modLoop) n
+    ]
+  , bgroup "IO"
+    [ bench "Church.StateC" $ whnfAppIO (C.Church.execState from . modLoop) n
+    , bench "Lazy.StateC"   $ whnfAppIO (C.Lazy.execState from . modLoop) n
+    , bench "Strict.StateC" $ whnfAppIO (C.Strict.execState from . modLoop) n
+    , bench "Lazy.StateT"   $ whnfAppIO (flip T.Lazy.execStateT from . modLoop) n
+    , bench "Strict.StateT" $ whnfAppIO (flip T.Strict.execStateT from . modLoop) n
+    ]
```

Comment:
Main takeaway: the Church-encoded `State` carrier is competitive with our strict `State` carrier in `Identity`, and a healthy margin faster in `IO`.